File: jvm/core/main/scala/com/webank/eggroll/core/containers/meta/ContainerContent.scala
package com.webank.eggroll.core.containers.meta

import com.google.protobuf.ByteString
import com.webank.eggroll.core.meta.Containers

import scala.language.implicitConversions

object ContainerContent {
  implicit def deserialize(byteString: ByteString): ContainerContent = {
    val src = Containers.ContainerContent.parseFrom(byteString)
    ContainerContent(
      containerId = src.getContainerId,
      content = src.getContent.toByteArray,
      compressMethod = src.getCompressMethod
    )
  }

  implicit def toProto(src: ContainerContent): Containers.ContainerContent = {
    val builder = Containers.ContainerContent.newBuilder()
      .setContainerId(src.containerId)
      .setContent(ByteString.copyFrom(src.content))
      .setCompressMethod(src.compressMethod)
    builder.build()
  }

  implicit def serialize(src: ContainerContent): Array[Byte] = {
    toProto(src).toByteArray
  }
}

case class ContainerContent(containerId: Long, content: Array[Byte], compressMethod: String)

File: jvm/core/src/main/java/org/fedai/eggroll/core/pojo/ContainerContent.java
package org.fedai.eggroll.core.pojo;

import com.google.protobuf.ByteString;
import com.webank.eggroll.core.meta.Containers;
import lombok.Data;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@Data
public class ContainerContent implements RpcMessage {

    Logger log = LoggerFactory.getLogger(ContainerContent.class);

    private int rank;
    private byte[] content;
    private String compressMethod;

    public ContainerContent() {

    }

    public ContainerContent(byte[] content, String compressMethod) {
        this.content = content;
        this.compressMethod = compressMethod;
    }

    public ContainerContent(int rank, byte[] content, String compressMethod) {
        this.rank = rank;
        this.content = content;
        this.compressMethod = compressMethod;
    }

    public Containers.ContainerContent toProto() {
        return Containers.ContainerContent.newBuilder()
//                .setRank(src.rank)
                .setContent(ByteString.copyFrom(content))
                .setCompressMethod(compressMethod)
                .build();
    }

    public ContainerContent fromProto(Containers.ContainerContent content) {
        ContainerContent containerContent = new ContainerContent();
        containerContent.deserialize(content.toByteArray());
        return containerContent;
    }

    @Override
    public byte[] serialize() {
        return toProto().toByteArray();
    }

    @Override
    public void deserialize(byte[] data) {
        try {
            Containers.ContainerContent src = Containers.ContainerContent.parseFrom(data);
            this.content = src.getContent().toByteArray();
            this.compressMethod = src.getCompressMethod();
        } catch (Exception e) {
            log.error("deserialize error : ", e);
        }
    }
}
